ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/V2/Init.hs

```diff
@@ -0,0 +1,377 @@
+        traceMarkerIO "Loaded snapshot"
+        pure s
+    , closeDb = closeLedgerSeq
+    , initReapplyBlock = \a b c -> do
+        (LedgerSeq x, y) <- reapplyThenPush lgrRegistry a b c
+        mapM_ (close . tables) (AS.toOldestFirst x)
+        pure y
+    , currentTip = ledgerState . current
+    , mkLedgerDb = \lseq -> do
+        traceMarkerIO "Initialize LedgerDB"
+        let (LedgerSeq rel, dbPrunedToImmDBTip) = pruneToImmTipOnly lseq
+        mapM_ (close . tables) (AS.toOldestFirst rel)
+        (varDB, prevApplied) <-
+          (,) <$> newTVarIO dbPrunedToImmDBTip <*> newTVarIO Set.empty
+        forkers <- newTVarIO Map.empty
+        nextForkerKey <- newTVarIO (ForkerKey 0)
+        lock <- RAWLock.new LDBLock
+        let env = LedgerDBEnv {
+                 ldbSeq            = varDB
+               , ldbPrevApplied    = prevApplied
+               , ldbForkers        = forkers
+               , ldbNextForkerKey  = nextForkerKey
+               , ldbSnapshotPolicy = defaultSnapshotPolicy (ledgerDbCfgSecParam lgrConfig) lgrSnapshotPolicyArgs
+               , ldbTracer         = lgrTracer
+               , ldbCfg            = lgrConfig
+               , ldbHasFS          = lgrHasFS
+               , ldbResolveBlock   = getBlock
+               , ldbQueryBatchSize = Nothing
+               , ldbReleaseLock    = AllowThunk lock
+               }
+        h <- LDBHandle <$> newTVarIO (LedgerDBOpen env)
+        pure $ implMkLedgerDb h bss
+    }
+ where
+   LedgerDbArgs {
+       lgrConfig
+     , lgrGenesis
+     , lgrHasFS
+     , lgrSnapshotPolicyArgs
+     , lgrTracer
+     , lgrRegistry
+     } = args
+
+   bss = case flavArgs of V2Args bss0 -> bss0
+
+   emptyF :: ExtLedgerState blk ValuesMK
+          -> m (LedgerSeq' m blk)
+   emptyF st =
+     empty' st $ case bss of
+                  InMemoryHandleArgs -> InMemory.newInMemoryLedgerTablesHandle lgrHasFS
+                  --TODO LSMHandleArgs      -> LSM.newLSMLedgerTablesHandle
+
+   loadSnapshot :: CodecConfig blk
+                -> SomeHasFS m
+                -> DiskSnapshot
+                -> m (Either (SnapshotFailure blk) (LedgerSeq' m blk, RealPoint blk))
+   loadSnapshot = case bss of
+     InMemoryHandleArgs -> InMemory.loadSnapshot lgrRegistry
+     --TODO LSMHandleArgs      -> LSM.loadSnapshot
+
+implMkLedgerDb ::
+     forall m l blk.
+     ( IOLike m
+     , HasCallStack
+     , IsLedger l
+     , l ~ ExtLedgerState blk
+     , StandardHash l, HasLedgerTables l
+#if __GLASGOW_HASKELL__ < 908
+     , HeaderHash l ~ HeaderHash blk
+#endif
+     , LedgerSupportsProtocol blk
+     , LedgerDbSerialiseConstraints blk
+     , MonadBase m m
+     , HasHardForkHistory blk
+     )
+  => LedgerDBHandle m l blk
+  -> HandleArgs
+  -> (LedgerDB m l blk, TestInternals m l blk)
+implMkLedgerDb h bss = (LedgerDB {
+      getVolatileTip            = getEnvSTM  h implGetVolatileTip
```

Comment:
Some of these `impl*` functions share a worrying amount of code with the V1 impl.

Reply:
We decided to now continue on this path as V1 will be deleted.